src/Telemetry/Tracks/class-tracks-event.php

```diff
@@ -0,0 +1,281 @@
+	 * @since 3.10.0
+	 *
+	 * @return bool|WP_Error True if the event is recordable.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function is_recordable() {
+		if ( ! Telemetry_System::is_backend_telemetry_allowed() ) {
+			return false;
+		}
+
+		// Don't record events during unit tests and CI runs.
+		if ( 'wptests_capabilities' === wp_get_current_user()->cap_key ) {
+			return false;
+		}
+
+		if ( is_wp_error( $this->data ) ) {
+			return $this->data;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Processes the event's properties to get them ready for validation.
+	 *
+	 * @since 3.10.0
+	 *
+	 * @param string                            $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any event properties to be processed.
+	 *
+	 * @return stdClass The resulting event object with processed properties.
+	 */
+	protected static function process_properties(
+		string $event_name,
+		array $event_properties
+	): stdClass {
+		$event = (object) self::sanitize_properties_array( $event_properties );
+		$event = self::set_user_properties( $event );
+
+		// Set event name.
+		$event->_en = preg_replace(
+			'/^(?:' . self::EVENT_NAME_PREFIX . ')?(.*)/',
+			self::EVENT_NAME_PREFIX . '\1',
+			$event_name
+		) ?? '';
+
+		// Set event timestamp.
+		if ( ! isset( $event->_ts ) ) {
+			$event->_ts = get_timestamp();
+		}
+
+		// Remove non-routable IPs to prevent record from being discarded.
+		if ( property_exists( $event, '_via_ip' ) &&
+			1 === preg_match( '/^192\.168|^10\./', $event->_via_ip ) ) {
+			unset( $event->_via_ip );
+		}
+
+		// Set VIP environment if it exists.
+		if ( defined( 'VIP_GO_APP_ENVIRONMENT' ) ) {
+			$app_environment = constant( 'VIP_GO_APP_ENVIRONMENT' );
+			if ( is_string( $app_environment ) && '' !== $app_environment ) {
+				$event->vipgo_env = $app_environment;
+			}
+		}
+
+		return $event;
+	}
+
+	/**
+	 * Sets the Tracks User ID and User ID Type depending on the current
+	 * environment.
+	 *
+	 * @since 3.10.0
+	 *
+	 * @param stdClass $event The event to annotate with identity information.
+	 *
+	 * @return stdClass The new event object including identity information.
+	 */
+	protected static function set_user_properties( stdClass $event ): stdClass {
+		$wp_user_id = get_current_user_id();
+
+		// Only track logged-in users.
+		if ( 0 === $wp_user_id ) {
+			return $event;
+		}
+
+		// Users in the VIP environment.
+		if ( defined( 'VIP_GO_APP_ID' ) ) {
+			$app_id = constant( 'VIP_GO_APP_ID' );
+			if ( is_integer( $app_id ) && 0 < $app_id ) {
+				$event->_ui = $app_id . '_' . $wp_user_id;
+				$event->_ut = 'vip_go_app_wp_user';
+
+				return $event;
+			}
+		}
+
+		// All other environments.
+		$home_option = get_option( 'home' );
+		if ( ! is_string( $home_option ) || '' === $home_option ) {
+			return $event;
+		}
+		$event->_ui = md5( "$home_option|$wp_user_id" );
```

Comment:
Use `wp_hash` (since it uses a salt) and fallback to site url.

```suggestion
		$wp_base_url = get_option( 'home' );
		if ( empty( $wp_base_url ) ) {
			$wp_base_url = get_option( 'siteurl' );
		}
		$event->_ui = wp_hash( sprintf( '%s|%s', $wp_base_url, $wp_user_id ) );
```